### DISCUSSION TITLE
When using Buildout, Greenlet is not properly installing as sub-dependency on GHA, macOS=>11 and Python 3.10

### DISCUSSION THREAD

**Discussion by amotl** [sqlalchemy/sqlalchemy#8182] [Usage Questions]
### Describe the bug

Dear @zzzeek, dear SQLAlchemy community.

### Introduction
This is not really an issue, but it is also not a usage question. It relates to #7922, but offers more general insights into similar problems when installing SQLAlchemy into a regular x86, non-M1 GHA environment with specific conditions.

_Please feel free to transfer it to the discussion space, or slap an `information` label on it and close the issue right away. It is merely meant to be filed as an anchor people can attach to and find a solution when running into similar problems._

So, we were tracing an issue which we reported about at https://github.com/crate-workbench/test-buildout-python310/issues/1 and https://github.com/actions/virtual-environments/issues/5791. I will try to give a short and concise summary about what we learned from others. In this regard, kudos to @mikhailkoliada.

### Summary
We learned that the `greenlet` package would not install on macOS runners on GHA environments, specifically on macOS>=11, but only when using Buildout.

It can be fixed by explicitly selecting the Xcode command line tools within the GHA workflow definition file, see https://github.com/crate-workbench/test-buildout-python310/pull/2:
```
sudo xcode-select -switch /Library/Developer/CommandLineTools
```
This paves the way that the build process can find the GCC compiler to build the C extension of the `greenlet` package. So far so good.

### Thoughts
However, the second part of our question, as summarized at https://github.com/actions/virtual-environments/issues/5791#issuecomment-1167829406, is still open: Why does the installation process not automatically select the `greenlet-1.1.2-cp310-cp310-macosx_10_14_x86_64.whl` binary wheel package for download and installation, which would not need any compiler to be addressed at all. Right?

With kind regards,
Andreas.

### To Reproduce

We have a dedicated reproduction repository over at [1]. Following your suggestion at [2]

> `pip install sqlalchemy[asyncio]` should definitely install greenlet in all cases.

the patch [3] explicitly does that, but also fails under the conditions specified above. Only @mikhailkoliada's suggestion at [4] repairs the build, but still needs to compile the C extension of `greenlet`.

[1] https://github.com/crate-workbench/test-buildout-python310
[2] https://github.com/sqlalchemy/sqlalchemy/issues/7922#issuecomment-1092811109
[3] https://github.com/crate-workbench/test-buildout-python310/pull/3
[4] https://github.com/crate-workbench/test-buildout-python310/pull/2

### Error

The broken build [1] contains stack traces from Buildout, which do not make sense in this context. On a fixed build like [2], you can see however that the `greenlet-1.1.2.tar.gz` sdist package gets selected for download instead of the corresponding binary wheel, which would be expected.

Remember it is only happening with GHA/Buildout on macOS>=11 and Python 3.10. On all other slots of the test matrix, the build succeeds by always selecting the binary wheel of the `greenlet` package.

[1] https://github.com/crate-workbench/test-buildout-python310/actions/runs/2571782179
[2] https://github.com/crate-workbench/test-buildout-python310/runs/7080203493#step:5:288

### Versions

- OS: macOS 11 and macOS 12
- Python: Python 3.10
- SQLAlchemy: SQLAlchemy 1.4

### Additional context

As said, this issue will probably not need any action. It is merely for informational purposes to support the community and carry some communication on other repositories back and forth.

**Comment by amotl** [comment]
As said, this issue will probably not need any action. It is merely for informational purposes to support the community and carry some communication on other repositories back and forth.

To summarize the solution to the problem outlined above once more:
1. If you need to keep Buildout, see https://github.com/crate-workbench/test-buildout-python310/pull/2.
2. If you can drop Buildout like https://github.com/crate/crate-python/pull/429, it will also resolve your problem.

We have chosen the second path.

**Reply by amotl** [reply]
P.S.: Thank you for moving this into a discussion, @zzzeek. Have a nice day!

**Comment by zzzeek** [comment]
> However, the second part of our question, as summarized at [actions/virtual-environments#5791 (comment)](https://github.com/actions/virtual-environments/issues/5791#issuecomment-1167829406), is still open: Why does the installation process not automatically select the `greenlet-1.1.2-cp310-cp310-macosx_10_14_x86_64.whl` binary wheel package for download and installation, which would not need any compiler to be addressed at all. Right?

So if you are on x86 and not M1, there should be nothing special happening here at all, as long as your hardware is reported into setup.cfg as `x86_64`.   We name the greenlet dependency as `greenlet`.   so installing SQLAlchemy would treat its greenlet dependency the same way as it would if you were to say `pip install greenlet` directly.    you should not have to install the `[asyncio]` target directly it should be automatic.

**Reply by amotl** [reply]
Dear Mike,

thank you for your quick response. I think I missed to include into my original post that I already communicated this with the people at the Buildout project at https://github.com/buildout/buildout/issues/609. Apologies.

> as long as your hardware is reported into `setup.cfg` as `x86_64`

I think this is specifically the problem here, exclusively in the context when using Buildout. See also https://github.com/buildout/buildout/issues/606, where @goschtl reported about his observations. The discussion between him and @gotcha makes me feel like this is the trace to the reason why the report to `setup.cfg` might be wrong.

With kind regards,
Andreas.
